rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+and update it on the backend, constantly keeping the dynamic apps state up to date. Writing to the backend will be handled
+by already existing mechanism in our code - reconciler. Reconciler will compare given fresh list of resources and resources we currently have
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name, like this:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+To allow for finer control over exposed service transformation into Teleport app Kubernetes annotations will be used.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
```

Comment:
Can we make auto-discovery a default for helm chart installs? @tigrato @r0mant @AntonAM ?

Reply:
@klizhentas 

We can as long as we update the documentation to mention that the token has to be valid for `kube,app, discovery`

Reply:
Even with updated documentation, this will break existing setups if the default configuration requires a new token type. We must gracefully disable app discovery if we don't have a cert with the discovery role. However, this behaviour can also be a footgun: instead of failing fast, the users end up with a half-working service.

Reply:
When I meant updating docs I meant updating all references to `tctl token create --types=kube,app,discovery`, `helm install .... --set roles=kube,app,discovery` so we don't break existing setups by starting services that we don't have a valid token for.

New installations will automatically enable the discovery service based on the roles provided. Having services shutting down silently isn't a good approach because it will cause a lot of support load.

Reply:
Yes, we should make it default for new installs but not for existing clusters.

Reply:
I'm not sure about making it default for helm chart. If we make it completely not requiring user setup it means we need to import all apps from the cluster and that sounds like it can potentially bring problems with security (overexposing apps) and confusion (users might not expect all the apps appearing in Teleport). Wouldn't it be better if users need to make conscious decision to do something like this? We could even show tip if they really want to import all apps (`namespace: ["*"];labesl: "*":"*"`). I like Tiago's idea for the Kubernetes Discover wizard with interactive UX.

Reply:
Right, for Discover we can design the UX we want.

I think for the standalone Helm chart it is also beneficial to have the discovery enabled by default. This is how people will get value quickly without reading a bunch of docs and configuring a bunch of things. `helm install teleport-kube-agent` and you have your apps in the cluster.

I agree that blindly importing every single service is not great. Can we come up with some heuristics and import things we're reasonably sure are applications? For example, if the port is 80 or 443 that's likely an app. Or if the appProtocol is set to "http" or "https". So our label matchers will be wildcard by default but they will import only things that look like applications.

False positives will probably happen but that should be an ok tradeoff for a much better first-time UX. WDYGT? @AntonAM @tigrato @klizhentas

Reply:
Yep, we can do heuristics, we actually already have it, when trying to understand right URI for the app (based both on appProtocol and port names), but that means we would leave out all tcp apps out there.
We probably can add optional matcher's field `ports` to be have something like this by default
```yaml
ports: ["http", "https", "80", "443"]
```
And we'll match it against port name or number.

Maybe instead of making app discovery default for helm chart, we add chart option `--set enable-app-discovery=true` on which we enable default discovery config with wildcards (with port filtering or not we can decide separately). So it's very easy to enable, users don't have to fiddle with configuring bunch of things, but they make conscious decision so at least they understand they will expose all apps. When users installing `teleport-kube-agent` they still need to provide a few options like kube cluster name and token, that can be just another simple option. What do you think?

Reply:
From the product UX perspective, I think it should be the opposite, enable it by default and let them opt out if they want. Otherwise they won't discover this option unless they know what they're looking for, (most) users don't read docs.

Reply:
OK, if we're making it default we just need to decide whether we want to process all services in the cluster or limit default to just what we heuristically think is a http app/port? (adding that `ports` config I described above).

Reply:
Yep, I think discovering what looks like an app using heuristics you described elsewhere in this RFD is a sane default.

Reply:
Ok, just to make sure and clarify things, doing a summary. 
So our intention is to discover "http-like" apps by default using our different heuristics (so we don't require explicit annotation for it). But we also want to let user import tcp apps if they want to, for that we'll require explicit annotation `teleport.dev/protocol: tcp`. Our heuristics for determining if it's a http/https app will include those already described in RFD (port names, numbers, `appProtocol` field) and also I've added port protocol checker as @tigrato suggested - it will send http request to check port if protocol is still unclear from other heuristics.
I'll also update RFD accordingly.